lib/FrontendTool/ReferenceDependencies.cpp

```diff
@@ -23,253 +23,308 @@
 #include "swift/AST/Types.h"
 #include "swift/Frontend/FrontendOptions.h"
+#include "swift/Frontend/ReferenceDependencyKeys.h"
 #include "swift/Basic/LLVM.h"
 #include "llvm/ADT/MapVector.h"
 #include "llvm/ADT/SetVector.h"
 #include "llvm/ADT/SmallVector.h"
 #include "llvm/Support/FileSystem.h"
 #include "llvm/Support/YAMLParser.h"
 
+// See docs/DependencyAnalysis.rst.
+
 using namespace swift;
 
-static void findNominalsAndOperators(
-    llvm::MapVector<const NominalTypeDecl *, bool> &foundNominals,
-    llvm::SmallVectorImpl<const FuncDecl *> &foundOperators,
-    DeclRange members) {
-  for (const Decl *D : members) {
-    auto *VD = dyn_cast<ValueDecl>(D);
-    if (!VD)
-      continue;
+std::vector<std::string>
+swift::reversePathSortedFilenames(const ArrayRef<std::string> elts) {
+  std::vector<std::string> tmp(elts.begin(), elts.end());
+  std::sort(tmp.begin(), tmp.end(),
+            [](const std::string &a, const std::string &b) -> bool {
+              return std::lexicographical_compare(a.rbegin(), a.rend(),
+                                                  b.rbegin(), b.rend());
+            });
+  return tmp;
+}
 
-    if (VD->hasAccess() &&
-        VD->getFormalAccess() <= AccessLevel::FilePrivate) {
-      continue;
-    }
+using namespace reference_dependency_keys;
 
-    if (VD->getFullName().isOperator()) {
-      foundOperators.push_back(cast<FuncDecl>(VD));
-      continue;
-    }
+static std::string escape(const DeclBaseName name) {
+  return llvm::yaml::escape(name.userFacingName());
+}
 
-    auto nominal = dyn_cast<NominalTypeDecl>(D);
-    if (!nominal)
-      continue;
-    foundNominals[nominal] |= true;
-    findNominalsAndOperators(foundNominals, foundOperators,
-                             nominal->getMembers());
-  }
+static std::string mangleTypeAsContext(const NominalTypeDecl *type) {
+  Mangle::ASTMangler Mangler;
+  return Mangler.mangleTypeAsContextUSR(type);
 }
 
-static bool declIsPrivate(const Decl *member) {
-  auto *VD = dyn_cast<ValueDecl>(member);
-  if (!VD) {
-    switch (member->getKind()) {
-    case DeclKind::Import:
-    case DeclKind::PatternBinding:
-    case DeclKind::EnumCase:
-    case DeclKind::TopLevelCode:
-    case DeclKind::IfConfig:
-    case DeclKind::PoundDiagnostic:
-      return true;
+template <typename Decl_T>
+static bool hasSomePrivateAccessLevel(const Decl_T *decl) {
+  return decl->hasAccess() &&
+         decl->getFormalAccess() <= AccessLevel::FilePrivate;
+}
 
-    case DeclKind::Extension:
-    case DeclKind::InfixOperator:
-    case DeclKind::PrefixOperator:
-    case DeclKind::PostfixOperator:
-      return false;
+/// If this declaration's access level confines it to a file, no other file can
+/// depend upon it.
+template <typename Decl_T>
+static bool isVisibleOutsideItsFile(const Decl_T *decl) {
+  return !hasSomePrivateAccessLevel(decl);
+}
 
-    default:
-      llvm_unreachable("everything else is a ValueDecl");
+class ReferenceDependencyFileEmitter {
```

Comment:
Nitpick: Please make sure these are in an [anonymous namespace](http://llvm.org/docs/CodingStandards.html#anonymous-namespaces), so that they don't generate public symbols.

Reply:
Thanks! That change pushed me into moving some functions around within the file a bit.